src/Middleware/ConcurrencyLimiter/test/PolicyTests/StackPolicyTests.cs

```diff
@@ -83,40 +94,49 @@ public static void ExitRequestsPreserveSemaphoreState()
         }));
 
         var task1 = stack.TryEnterAsync();
-        Assert.True(task1.IsCompleted && task1.Result);
+        Assert.True(task1.IsCompleted && await task1);
 
         var task2 = stack.TryEnterAsync();
         Assert.False(task2.IsCompleted);
 
         stack.OnExit();  // t1 exits, should free t2 to return
-        Assert.True(task2.IsCompleted && task2.Result);
+        Assert.True(await task2);
 
         stack.OnExit();  // t2 exists, there's now a free spot in server
 
         var task3 = stack.TryEnterAsync();
-        Assert.True(task3.IsCompleted && task3.Result);
+        Assert.True(task3.IsCompleted && await task3);
     }
 
     [Fact]
-    public static void StaleRequestsAreProperlyOverwritten()
+    public async Task StaleRequestsAreProperlyOverwritten()
     {
         var stack = new StackPolicy(Options.Create(new QueuePolicyOptions
         {
-            MaxConcurrentRequests = 0,
+            MaxConcurrentRequests = 1,
             RequestQueueLimit = 4,
         }));
 
+        var task0 = stack.TryEnterAsync();
+        Assert.True(task0.IsCompleted && await task0);
+
         var task1 = stack.TryEnterAsync();
+        Assert.False(task1.IsCompleted);
+
         stack.OnExit();
-        Assert.True(task1.IsCompleted);
+        Assert.True(await task1);
```

Comment:
Why is it `Assert.True(task0.IsCompleted && await task0)` above but just `Assert.True(await task1);` here? Is `task1` not completed immediately?

Reply:
There's a race where OnExit causes task1's completion to get dispatched rather than completed inline.

Reply:
I see. It looks like it's because ConcurrencyLimiter uses `TaskCreationOptions.RunContinuationsAsynchronously` for the TCS. If we could look at the `TCS.Task` directly, it should be completed, but that's not the case here.